core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -89,6 +89,7 @@
   private final List<ManifestFile> rewrittenAppendManifests = Lists.newArrayList();
   private final SnapshotSummary.Builder addedFilesSummary = SnapshotSummary.builder();
   private final SnapshotSummary.Builder appendedManifestsSummary = SnapshotSummary.builder();
+  private final Set<StructLike> changedPartitions = Sets.newHashSet();
```

Comment:
Can we use PartitionSet ?

Reply:
That's a very good point.

We should either use `PartitionSet` if we need to have partitions per spec or just use `StructLikeSet`. We cannot use a regular set as we may get arbitrary StructLike implementations. Most likely, we will need `PartitionSet`. Can we add a test with multiple specs?

Reply:
I think we should even add a style rule to prohibit regular sets/maps for `StructLike` in a separate PR.